### ISSUE TITLE
Grain.AsReference Public API change suggestion

### ISSUE THREAD

**Issue by gabikliot** [dotnet/orleans#195]
Grain base class currently has an extension method `AsReference` that returns a `GrainReference` to this grain:
`public static GrainReference AsReference(this Runtime.IAddressable grain)`
The usage patterns is for a grain to get a `GrainReference` to self, Cast it to an actual interface that it implements and the use it (either pass to someone else or store in storage).
`public static TGrainInterface Cast<TGrainInterface>(this Runtime.IAddressable grain)`

The problem is that the `GrainReference` it returns is weakly typed. That is, this GR is not bound to any particular interface that this grain implements (kind of like "void*"). The implementation of `AsReference` cannot bind it to the interface itself, since in the general case grain can implement more than one interface. 
If the application code always makes sure to Cast the weakly-typed `GrainReference` to one of the interfaces, there is no problem., However, in the general case an application may use this GR without casting it. As a result we may end up with "half-baked" weakly-typed GR leaking into the wild.

The proposal is to change the `AsReference` to always return a strongly typed GR:
`public static TGrainInterface AsReference<TGrainInterface>(this Runtime.IAddressable grain)`
basically unifying `AsReference` and `Cast` into one method.
That way we will eliminate any public API to get a weakly-typed GR.

We are seeking community feedback on this change.

**Comment by veikkoeeva** [comment]
I haven't used this feature, so my opinion isn't based on actual usage. It follows I don't have backwards compatibility problems (and would they be a real problem?) and I prefer stronger typing.

**Comment by jkonecki** [comment]
I fully agree with Veikko.

**Comment by sergeybykov** [comment]
This change should also help resolve https://github.com/dotnet/orleans/issues/71, as Cast() shouldn't be needed anymore.

**Comment by gabikliot** [comment]
Fixed via #288. Closing.
